Add new `QueuedJobRegistry` to catch jobs dropped by workers before m…

joshcoden opened the pull request:
…arked as started

Fixes #1553

selwin commented:
Thanks for the PR.

I think at any one time, a job can only exist in one registry. When a job is moved to `StartedJobRegistry`, it should also be removed from `QueuedJobRegistry` in a single pipeline call. Do you mind making this change?

selwin commented on docs/docs/job_registries.md:
```diff
@@ -12,6 +12,8 @@ executed and removed right after completion (success or failure).
 job to finish).
 * `ScheduledJobRegistry` Holds scheduled jobs.
 * `CanceledJobRegistry` Holds canceled jobs.
+* `QueuedJobRegistry` Holds jobs from the time they are enqueued, until they are moved to the `QueuedJobRegistry`.
```
I think this should be `until they are moved to the `StartedJobRegistry``

joshcoden commented:
> > It is when you enqueue, it's just up a level in the call stack:
> 
> I meant when a job is popped off the queue and moved to `StartedJobRegistry` (when it's being worked on), it should also be removed from `QueuedJobRegistry`. As it is now, when a job is being worked on it exists in both `QueuedJobRegistry` and `StartedJobRegistry`

@selwin as mentioned above, I changed it to remove it, please see the code link in the comment above:  https://github.com/rq/rq/pull/1568#issuecomment-928559454: 
https://github.com/rq/rq/blob/8fd8de0742490b5d0fb33a9e9564a0ff609d52fe/rq/worker.py#L913

alella commented:
| fn_call                      | queue | QueuedJobRegistry | StartedJobRegistry | parent worker | forked worker |
|------------------------------|-------|-------------------|--------------------|---------------|---------------|
| enqueue                      | job   | job               |                    | -             | -             |
| dequeue_job_and_maintain_ttl | -     | job               |                    | active        | doesnt exist  |
| perform_job(job, queue)      | -     | job               |                    | inactive      | active        |
| [prepare_job_execution(job)](https://github.com/rq/rq/blob/8fd8de0742490b5d0fb33a9e9564a0ff609d52fe/rq/worker.py#L913-L914)   | -     | -                 | job                | inactive      | active        |

I think what @joshcoden is trying to say is prepare_job_execution calls `job.queued_job_registry.remove` and `job.heartbeat` in the same pipeline. [job.heartbeat](https://github.com/rq/rq/blob/8fd8de0742490b5d0fb33a9e9564a0ff609d52fe/rq/job.py#L445-L449) in turn calls `started_job_registry.add` using the same pipeline. So the job never exists in both `QueuedJobRegistry` and `StartedJobRegistry` at the same time

joshcoden commented on rq/queue.py:
```diff
@@ -560,11 +566,13 @@ def enqueue_job(self, job, pipeline=None, at_front=False):
         job.origin = self.name
         job.enqueued_at = utcnow()
-
+        # TODO: Specify xx operator to prev'tent jobs from being double added to registry
```
@alella  @selwin Should `xx` just be set to `ttl`?  I guess it would just update the score if it already exists depending on which other arguments we pass in.

joshcoden commented on rq/registry.py:
```diff
@@ -334,6 +333,72 @@ def cleanup(self):
+                        serializer=self.serializer
+                    )
+                    if job.enqueued_at:
+                        if job.enqueued_at < front_timestamp:
+                            self.requeue(job)
```
@alella The user can manually specify the score when calling `add` on the registry.  While no such calls will exist in code, a user code specify a `ttl` that does not align with the `enqueued_at` time:

https://github.com/rq/rq/blob/47110806d1e281f7ba314f837658e247eaaf7a88/rq/registry.py#L64-L72

joshcoden commented on rq/registry.py:
```diff
@@ -334,6 +333,72 @@ def cleanup(self):
+                        if job.enqueued_at < front_timestamp:
+                            self.requeue(job)
+                    else:
+                        # TODO: Should we do anything else?
```
@alella My only other thought would be to log an error and at least try other jobs, but this should really never happen so it's probably safer to just raise in this case.

selwin commented:
@alella @joshcoden thanks for explaination.

It is indeed working as expected. I misread this because the the code around this part is not symmetric. Can you create a `job.move_to_started_job_registry()` method that moves the job from `QueuedJobRegistry` into `StartedJobRegistry`? It will be cleaner this way.

With this, we can potentially also cleanup `job.heartbeat(xx)` because right now it does two jobs:
1. Adding job to started job registry
2. Updating the timestamp in started job registry

This already led to race condition bugs fixed in this PR https://github.com/rq/rq/pull/1550

selwin commented on rq/registry.py:
```diff
@@ -334,6 +333,66 @@ def cleanup(self):
+            front_job = jobs[0]
+            if front_job.enqueued_at is not None:
+                return front_job.enqueued_at
+            else:
+                raise InvalidJobOperationError(
+                    "Queued job {} has no enqueue_at value!".format(front_job.id)
+                )
```
This could not happen, right? So I think we can skip this check.

selwin commented on rq/registry.py:
```diff
@@ -334,6 +333,66 @@ def cleanup(self):
+        for job_id in job_ids:
+            # If job was enqueued AFTER the front of the queue it must have already been dequeued
+            # This is faster than seeing if the job is in the queue directly.
+            front_timestamp = self._get_front_queue_timestamp() 
```
We can move this out of the loop so we don't keep getting new timestamp for every single job processed.

selwin commented on rq/registry.py:
```diff
@@ -334,6 +333,66 @@ def cleanup(self):
+        It is not defined in cleanup since we don't want this being called everytime count or get_job_ids is called
+        """
+        job_ids = self.get_job_ids()
+        for job_id in job_ids:
```
This could be expensive if we have lots of jobs in `QueuedJobRegistry` and sometimes people have millions of jobs enqueued. I wonder if we have a way to only check a subset of the jobs.

selwin commented on rq/registry.py:
```diff
@@ -352,3 +411,9 @@ def clean_registries(queue):
+                                 connection=queue.connection,
+                                 job_class=queue.job_class,
+                                 serializer=queue.serializer)
+    registry.requeue_stuck_jobs()
```
I'm not sure if we should automatically requeue stuck jobs because:
1. This operation could be heavy depending on the number of jobs you have in the registry
2. If requeued, jobs that are time sensitive could lead to unwanted outcome

lowercase00 commented:
@joshcoden any plans to continue working on this?
